runners/local-artifact-service-java/src/main/java/org/apache/beam/artifact/local/LocalFileSystemArtifactRetrievalService.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.artifact.local;
+
+import com.google.protobuf.ByteString;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.GetManifestResponse;
+import org.apache.beam.model.jobmanagement.v1.ArtifactRetrievalServiceGrpc;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactRetrievalService;
+
+/** An {@code ArtifactRetrievalService} which stages files to a local temp directory. */
+public class LocalFileSystemArtifactRetrievalService
+    extends ArtifactRetrievalServiceGrpc.ArtifactRetrievalServiceImplBase
+    implements ArtifactRetrievalService {
+  private static final long DEFAULT_CHUNK_SIZE = 2 * 1024 * 1024;
+
+  public static LocalFileSystemArtifactRetrievalService forRootDirectory(File base) {
+    return new LocalFileSystemArtifactRetrievalService(base);
+  }
+
+  private final LocalArtifactStagingLocation location;
+
+  private LocalFileSystemArtifactRetrievalService(File rootDirectory) {
+    this.location = LocalArtifactStagingLocation.forExistingDirectory(rootDirectory);
+  }
+
+  /**
+   * Get the manifest of artifacts that can be retrieved by this {@link
+   * LocalFileSystemArtifactRetrievalService}.
+   */
+  private ArtifactApi.Manifest getManifest() throws IOException {
+    return ArtifactApi.Manifest.parseFrom(new FileInputStream(location.getManifestFile()));
+  }
+
+  @Override
+  public final void getManifest(
+      ArtifactApi.GetManifestRequest request,
+      StreamObserver<GetManifestResponse> responseObserver) {
+    try {
+      responseObserver.onNext(GetManifestResponse.newBuilder().setManifest(getManifest()).build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  /** Get the artifact with the provided name as a sequence of bytes. */
+  private Iterable<ByteBuffer> getArtifact(String name) throws IOException {
+    File artifact = location.getArtifactFile(name);
+    if (!artifact.exists()) {
+      throw new FileNotFoundException(String.format("No such artifact %s", name));
+    }
+    FileChannel input = new FileInputStream(artifact).getChannel();
+    List<ByteBuffer> chunks = new ArrayList<>();
+    do {
+      ByteBuffer buf = ByteBuffer.allocate((int) Math.min(DEFAULT_CHUNK_SIZE, input.size()));
+      input.read(buf);
+      buf.flip();
+      chunks.add(buf);
+    } while (chunks.size() * DEFAULT_CHUNK_SIZE < input.size());
```

Comment:
This will load everything into memory. Consider using https://docs.oracle.com/javase/8/docs/api/java/nio/channels/FileChannel.html#map-java.nio.channels.FileChannel.MapMode-long-long- and moving the reading part into the getArtifact loop.

Reply:
Done.